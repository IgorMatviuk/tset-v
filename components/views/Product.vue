<template>
  <main class="product-page">
    <TheBreadcrumb :items="breadcrumb" />
    <BlockProduct :product="product" :links="links" />
    <BlockProductsList
      v-if="sections.similar_products.length"
      :products="sections.similar_products"
      :heading="$t('productPage.similarProducts')"
      parent-class="product-page__similar-products"
      class="product-page__similar-products"
    />
    <BlockProductsList
      :products="sections.featured_products"
      :heading="$t('productPage.featuredProducts')"
      parent-class="product-page__bestsellers"
      class="product-page__bestsellers"
    />
    <client-only>
      <BlockProductsList
        v-if="viewedProducts.length"
        :products="viewedProducts"
        :heading="$t('productPage.viewedProducts')"
        parent-class="product-page__viewed-products"
        class="product-page__viewed-products"
      />
    </client-only>
    <BlockSiteDescription :content="seoText" />
    <TheImageViewer
      v-if="isImageViewerVisible"
      :active-index="galleryActiveIndex"
    />
  </main>
</template>

<script>
import { createNamespacedHelpers } from 'vuex'

import TheBreadcrumb from '@/components/layout/breadcrumb/TheBreadcrumb'
import BlockProduct from '@/components/blocks/BlockProduct'
import BlockProductsList from '@/components/blocks/BlockProductsList'
import BlockSiteDescription from '@/components/blocks/BlockSiteDescription'
import TheImageViewer from '@/components/layout/image-viewer/TheImageViewer'

import breadrumb from '@/mixins/breadrumb'

const { mapState } = createNamespacedHelpers('product')

export default {
  components: {
    TheBreadcrumb,
    BlockProduct,
    BlockProductsList,
    BlockSiteDescription,
    TheImageViewer
  },
  mixins: [breadrumb()],
  data() {
    return {
      viewedProducts: []
    }
  },
  computed: {
    ...mapState({
      product: (state) => state.productData,
      links: (state) => state.links,
      sections: (state) => state.sectionsData,
      isImageViewerVisible: (state) => state.imageViewerVisibility
    }),
    seoText() {
      return this.$store.getters['product/seoText']
    },
    galleryActiveIndex() {
      const formattedPreviews = this.$store.state.product.activeVariant.previews.filter(
        (preview) => preview.preview_type !== 'video'
      )
      const index = formattedPreviews.findIndex(
        (item) => item === this.$store.state.product.activePreview
      )

      return index >= 0 ? index : 0
    }
  },
  async mounted() {
    // TODO: temporary solution since there is no any auth token logic built
    // The auth tokens will be implemented in the future release
    await this.$viewedProducts.connectToDatabase()
    await this.$viewedProducts.incrementing(this.product)

    this.viewedProducts = await this.$viewedProducts.getAll()
  }
}
</script>

<style lang="scss"></style>
