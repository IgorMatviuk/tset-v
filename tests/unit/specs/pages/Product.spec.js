import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'

import Product from '@/components/views/Product'

import {
  mutationTypes as rootMutationTypes,
  mutations as rootMutations
} from '@/store'

import {
  mutations as productMutations,
  mutationTypes as productMutationsTypes,
  getters as productGetters
} from '@/store/product'

import createFactory from '@/api/helpers/factory'
import productAttrs from '@/api/factories/product'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Product', () => {
  let store
  let mockViewedProducts
  let activeVariant
  let nextVariant
  let activePreview
  let nextPreview

  function createProduct() {
    const wrapper = shallowMount(Product, {
      localVue,
      store,
      mocks: {
        $t: (str) => str,
        $i18n: {
          path: (str) => str
        },
        $viewedProducts: mockViewedProducts
      },
      stubs: ['client-only']
    })

    return wrapper
  }

  beforeEach(() => {
    mockViewedProducts = {
      connectToDatabase: jest.fn(),
      incrementing: jest.fn(),
      getAll: jest.fn().mockResolvedValue([])
    }

    const products = createFactory(productAttrs, 2)

    activeVariant = products[0].variants[0]
    nextVariant = products[1].variants[0]
    activePreview = activeVariant.previews[3]
    nextPreview = nextVariant.previews[2]

    store = new Vuex.Store({
      state: {
        locale: 'ru',
        locales: ['uk', 'ru'],
        breadcrumb: []
      },
      mutations: { ...rootMutations },
      modules: {
        product: {
          namespaced: true,
          state: {
            productData: {
              attribute_family: {
                description: {
                  description: {
                    value: '<p>Product description</p>'
                  }
                }
              }
            },
            activeVariant,
            sectionsData: {
              featured_products: [],
              similar_products: [],
              viewed_products: []
            },
            links: {
              prev: 'prev-product-slug',
              next: 'next-product-slug'
            },
            imageViewerVisibility: true,
            activePreview
          },
          mutations: { ...productMutations },
          getters: { ...productGetters }
        }
      }
    })
  })

  test('renders component itself (RU)', () => {
    const wrapper = createProduct()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders component itself (UA)', () => {
    store.commit(rootMutationTypes.SET_LANG, 'uk')

    const wrapper = createProduct()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders components with default gallery active index', () => {
    const wrapper = createProduct()

    wrapper.vm.$store.commit(
      `product/${productMutationsTypes.SET_ACTIVE_PREVIEW}`,
      nextPreview
    )

    expect(wrapper.element).toMatchSnapshot()
  })
})
