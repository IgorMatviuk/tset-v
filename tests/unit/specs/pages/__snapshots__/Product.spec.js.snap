// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders component itself (RU) 1`] = `
<main
  class="product-page"
>
  <thebreadcrumb-stub
    items="[object Object]"
  />
   
  <blockproduct-stub
    links="[object Object]"
    product="[object Object]"
  />
   
  <!---->
   
  <blockproductslist-stub
    class="product-page__bestsellers"
    heading="productPage.featuredProducts"
    parentclass="product-page__bestsellers"
    products=""
  />
   
  <client-only-stub>
    <!---->
  </client-only-stub>
   
  <blocksitedescription-stub
    content=<p>Product description</p>
  />
   
  <theimageviewer-stub
    activeindex="3"
  />
</main>
`;

exports[`Product renders component itself (UA) 1`] = `
<main
  class="product-page"
>
  <thebreadcrumb-stub
    items="[object Object]"
  />
   
  <blockproduct-stub
    links="[object Object]"
    product="[object Object]"
  />
   
  <!---->
   
  <blockproductslist-stub
    class="product-page__bestsellers"
    heading="productPage.featuredProducts"
    parentclass="product-page__bestsellers"
    products=""
  />
   
  <client-only-stub>
    <!---->
  </client-only-stub>
   
  <blocksitedescription-stub
    content=<p>Product description</p>
  />
   
  <theimageviewer-stub
    activeindex="3"
  />
</main>
`;

exports[`Product renders components with default gallery active index 1`] = `
<main
  class="product-page"
>
  <thebreadcrumb-stub
    items="[object Object]"
  />
   
  <blockproduct-stub
    links="[object Object]"
    product="[object Object]"
  />
   
  <!---->
   
  <blockproductslist-stub
    class="product-page__bestsellers"
    heading="productPage.featuredProducts"
    parentclass="product-page__bestsellers"
    products=""
  />
   
  <client-only-stub>
    <!---->
  </client-only-stub>
   
  <blocksitedescription-stub
    content=<p>Product description</p>
  />
   
  <theimageviewer-stub
    activeindex="3"
  />
</main>
`;
